plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/AddFilesProcedure.java

```diff
@@ -0,0 +1,403 @@
+            }
+            if (!"MANAGED_TABLE".equalsIgnoreCase(sourceTable.getTableType()) && !"EXTERNAL_TABLE".equalsIgnoreCase(sourceTable.getTableType())) {
+                throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support adding files from %s table type".formatted(sourceTable.getTableType()));
+            }
+            if (isDeltaLakeTable(sourceTable)) {
+                throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support adding files from Delta Lake tables");
+            }
+            if (isHudiTable(sourceTable)) {
+                throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support adding files from Hudi tables");
+            }
+            if (isIcebergTable(sourceTable)) {
+                throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support adding files from Iceberg tables");
+            }
+            if (sourceTable.getPartitionColumns().isEmpty() && partitionFilter != null && !partitionFilter.isEmpty()) {
+                throw new TrinoException(NOT_SUPPORTED, "Partition filter is not supported for non-partitioned tables");
+            }
+            storageFormat = extractHiveStorageFormat(sourceTable.getStorage().getStorageFormat());
+            sourceLocation = sourceTable.getStorage().getLocation();
+            sourceType = SourceType.TABLE;
+        }
+        else if (location != null && format != null) {
+            checkProcedureArgument(sourceSchemaName == null, "SOURCE_SCHEMA_NAME argument must be null");
+            checkProcedureArgument(sourceTableName == null, "SOURCE_TABLE_NAME argument must be null");
+            checkProcedureArgument(partitionFilter == null, "PARTITION_FILTER argument must be null");
+
+            sourceLocation = location;
+            storageFormat = HiveStorageFormat.valueOf(format.toUpperCase(ENGLISH));
+            checkProcedureArgument(storageFormat == ORC || storageFormat == PARQUET || storageFormat == AVRO, "The procedure does not support storage format: %s", storageFormat);
+            sourceType = SourceType.LOCATION;
+        }
+        else {
+            throw new TrinoException(INVALID_PROCEDURE_ARGUMENT, "Either source schema and table names, or location and format must be provided");
+        }
+
+        try {
+            TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+            ImmutableList.Builder<DataFile> dataFilesBuilder = ImmutableList.builder();
+            if (partitionSpec.isUnpartitioned()) {
+                log.debug("Building data files from %s", sourceLocation);
+                dataFilesBuilder.addAll(buildDataFiles(fileSystem, sourceType, recursive, storageFormat, sourceLocation, partitionSpec, Optional.empty(), schema, requiredFields));
+            }
+            else {
+                checkProcedureArgument(location == null, "The procedure does not support partitioned tables with location argument");
```

Comment:
Can this check be moved above, under `else if (location != null && format != null) `. All validation happens there.